lib/Twig/Extension/Core.php

```diff
@@ -821,10 +821,32 @@ function _twig_default_filter($value, $default = '')
 function twig_get_array_keys_filter($array)
 {
     if ($array instanceof Traversable) {
-        return array_keys(iterator_to_array($array));
+        if ($array instanceof IteratorAggregate) {
+            $array = $array->getIterator();
+        } // else $array must be an instance of IteratorAggregate here, we could call getIterator on that one etc. until we get an iterator
+
+        if ($array instanceof Iterator) {
```

Comment:
the main advantage here is not accessing the objects, so classes implementing Iterator that use lazy loading of the objects are not triggered to do so on key request only